### PR TITLE
Translation into Russian. Full translation of chapter 2.

### PR DESCRIPTION
**Pull request by blademoon**
Finished the full Russian translation of chapter 2. Including missing (not translated before) files.  #75

### CODE REVIEW

**Comment by blademoon** [comment]
@MKhalusova Good day Maria. Can you please do a review?

File: chapters/ru/chapter2/3.mdx [hunk 1]
```diff
@@ -99,7 +99,7 @@ from transformers import BertConfig, BertModel
 config = BertConfig()
 model = BertModel(config)
 
-# Модель инициализируется случайным образом!
+# Модель инициализируется случайным образом!!
```
Comment (MKhalusova) [inline chapters/ru/chapter2/3.mdx:102]:
```suggestion
# Модель инициализируется случайным образом!
```

File: chapters/ru/chapter2/4.mdx [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+
+Однако модели могут обрабатывать только числа, поэтому нам нужно найти способ преобразовать исходный текст в числа. Этим занимаются токенизаторы, и существует множество способов сделать это. Цель состоит в том, чтобы найти наиболее осмысленное представление - то есть то, которое имеет наибольший смысл для модели, - и, если возможно, наименьшее представление.
+
+Давайте рассмотрим несколько примеров алгоритмов токенизации и постараемся ответить на некоторые вопросы, которые могут у вас возникнуть по токенизации.
+
+## На основе слов[[word-based]]
+
+<Youtube id="nhJxYji1aho"/>
+
+Первый тип токенайзера, который приходит на ум, - это _на основе слов (word-based)_. Как правило, ее очень легко настроить и использовать с помощью всего нескольких правил, и она часто дает достойные результаты. Например, на изображении ниже цель состоит в том, чтобы разбить исходный текст на слова и найти для каждого из них числовое представление:
```
Comment (MKhalusova) [inline chapters/ru/chapter2/4.mdx:43]:
```suggestion
Первый тип токенайзера, который приходит на ум, - это _на основе слов (word-based)_. Как правило, его очень легко настроить и использовать с помощью всего нескольких правил, и он часто дает достойные результаты. Например, на изображении ниже цель состоит в том, чтобы разбить исходный текст на слова и найти для каждого из них числовое представление:
```

File: chapters/ru/chapter2/4.mdx [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+
+<div class="flex justify-center">
+  <img class="block dark:hidden" src="https://huggingface.co/datasets/huggingface-course/documentation-images/resolve/main/en/chapter2/word_based_tokenization.svg" alt="An example of word-based tokenization."/>
```
Comment (MKhalusova) [inline chapters/ru/chapter2/4.mdx:46]: Please translate the alt text for the image

File: chapters/ru/chapter2/4.mdx [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+  <img class="hidden dark:block" src="https://huggingface.co/datasets/huggingface-course/documentation-images/resolve/main/en/chapter2/word_based_tokenization-dark.svg" alt="An example of word-based tokenization."/>
```
Comment (MKhalusova) [inline chapters/ru/chapter2/4.mdx:47]: Please translate the alt text for the image

File: chapters/ru/chapter2/4.mdx [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+
+Токенизаторы на основе символов (character-based) разбивают текст на символы, а не на слова. Это дает два основных преимущества:
+
+- Словарь намного меньше.
+- Неизвестных токенов гораздо меньше, поскольку каждое слово может быть образовано из символов.
+
+Но и здесь возникают некоторые вопросы, связанные с пробелами и пунктуацией:
+
+<div class="flex justify-center">
+  <img class="block dark:hidden" src="https://huggingface.co/datasets/huggingface-course/documentation-images/resolve/main/en/chapter2/character_based_tokenization.svg" alt="An example of character-based tokenization."/>
```
Comment (MKhalusova) [inline chapters/ru/chapter2/4.mdx:83]: Please translate the alt text for the image

File: chapters/ru/chapter2/4.mdx [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+  <img class="hidden dark:block" src="https://huggingface.co/datasets/huggingface-course/documentation-images/resolve/main/en/chapter2/character_based_tokenization-dark.svg" alt="An example of character-based tokenization."/>
```
Comment (MKhalusova) [inline chapters/ru/chapter2/4.mdx:84]: Please translate the alt text for the image

File: chapters/ru/chapter2/4.mdx [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+<Youtube id="zHvTiHr506c"/>
+
+Алгоритмы токенизации подслов (subword tokenization) основываются на принципе, согласно которому часто используемые слова не должны разбиваться на более мелкие подслова, а редкие слова должны быть разложены на значимые подслова.
+
+Например, "annoyingly" может считаться редким словом и может быть разложено на "annoying" и "ly". Оба они, скорее всего, будут чаще появляться как самостоятельные подслова, но в то же время значение " annoyingly" сохраняется за счет составного значения "annoying" и "ly".
+
+Вот пример, показывающий, как алгоритм токенизации подслов будет токенизировать последовательность "Let's do tokenization!":
+
+<div class="flex justify-center">
+  <img class="block dark:hidden" src="https://huggingface.co/datasets/huggingface-course/documentation-images/resolve/main/en/chapter2/bpe_subword.svg" alt="A subword tokenization algorithm."/>
```
Comment (MKhalusova) [inline chapters/ru/chapter2/4.mdx:104]: Please translate the alt text for the image

File: chapters/ru/chapter2/4.mdx [hunk 8]
```diff
@@ -0,0 +1,240 @@
...
+  <img class="hidden dark:block" src="https://huggingface.co/datasets/huggingface-course/documentation-images/resolve/main/en/chapter2/bpe_subword-dark.svg" alt="A subword tokenization algorithm."/>
```
Comment (MKhalusova) [inline chapters/ru/chapter2/4.mdx:105]: Please translate the alt text for the image

File: chapters/ru/chapter2/4.mdx [hunk 9]
```diff
@@ -0,0 +1,240 @@
...
+ 'attention_mask': [1, 1, 1, 1, 1, 1, 1, 1, 1]}
+```
+
+Сохранение токенизатора идентично сохранению модели:
+
+```py
+tokenizer.save_pretrained("directory_on_my_computer")
+```
+
+Подробнее о `token_type_ids` мы поговорим в [Главе 3](../chapter3), а ключ `attention_mask` мы объясним чуть позже. Сначала давайте посмотрим, как генерируются `input_ids`. Для этого нам понадобится рассмотреть промежуточные методы токенизатора.
```
Comment (MKhalusova) [inline chapters/ru/chapter2/4.mdx:166]:
```suggestion
Подробнее о `token_type_ids` мы поговорим в [Главе 3](../chapter3/1), а ключ `attention_mask` мы объясним чуть позже. Сначала давайте посмотрим, как генерируются `input_ids`. Для этого нам понадобится рассмотреть промежуточные методы токенизатора.
```

File: chapters/ru/chapter2/5.mdx [hunk 10]
```diff
@@ -0,0 +1,338 @@
+<FrameworkSwitchCourse {fw} />
+
+# Обработка множественных последовательностей[[handling-multiple-sequences]]
```
Comment (MKhalusova) [inline chapters/ru/chapter2/5.mdx:3]:
```suggestion
# Обработка нескольких последовательностей[[handling-multiple-sequences]]
```

File: chapters/ru/chapter2/5.mdx [hunk 11]
```diff
@@ -0,0 +1,338 @@
...
+
+{#if fw === 'pt'}
+<Youtube id="M6adb1j2jPI"/>
+{:else}
+<Youtube id="ROxrFOEbsQE"/>
+{/if}
+
+В предыдущем разделе мы рассмотрели самый простой вариант использования: проведение инференса на одной последовательности небольшой длины. Однако уже сейчас возникают некоторые вопросы:
+
+- Как нам работать с множественными последовательностями?
```
Comment (MKhalusova) [inline chapters/ru/chapter2/5.mdx:33]:
```suggestion
- Как нам работать с несколькими последовательностями?
```

File: chapters/ru/chapter2/5.mdx [hunk 12]
```diff
@@ -0,0 +1,338 @@
...
+checkpoint = "distilbert-base-uncased-finetuned-sst-2-english"
+tokenizer = AutoTokenizer.from_pretrained(checkpoint)
+model = AutoModelForSequenceClassification.from_pretrained(checkpoint)
+
+sequence = "I've been waiting for a HuggingFace course my whole life."
+
+tokens = tokenizer.tokenize(sequence)
+ids = tokenizer.convert_tokens_to_ids(tokens)
+input_ids = torch.tensor(ids)
+# Эта строка завершится неудачей.
```
Comment (MKhalusova) [inline chapters/ru/chapter2/5.mdx:58]:
```suggestion
# Эта строка выдаст ошибку.
```

File: chapters/ru/chapter2/5.mdx [hunk 13]
```diff
@@ -0,0 +1,338 @@
...
+checkpoint = "distilbert-base-uncased-finetuned-sst-2-english"
+tokenizer = AutoTokenizer.from_pretrained(checkpoint)
+model = TFAutoModelForSequenceClassification.from_pretrained(checkpoint)
+
+sequence = "I've been waiting for a HuggingFace course my whole life."
+
+tokens = tokenizer.tokenize(sequence)
+ids = tokenizer.convert_tokens_to_ids(tokens)
+input_ids = tf.constant(ids)
+# Эта строка завершится неудачей.
```
Comment (MKhalusova) [inline chapters/ru/chapter2/5.mdx:79]:
```suggestion
# Эта строка выдаст ошибку.
```

**Review by MKhalusova — COMMENTED** [review]
Great work on translation! I left a few suggestions.

File: chapters/ru/chapter2/1.mdx [hunk 14]
```diff
@@ -1,24 +1,25 @@
...
+Для решения этой проблемы была создана библиотека 🤗 Transformers. Ее цель - предоставить единый API, с помощью которого можно загрузить, обучить и сохранить любую модель Transformer. Основными особенностями библиотеки являются:
 
-- **Удобство в использовании**: Скачивание, загрузку и использование современной модели NLP для вывода данных, можно выполнять всего двумя строками кода.
-- **Гибкость**: По своей сути все модели представляют собой простые классы библиотек PyTorch `nn.Module` или TensorFlow `tf.keras.Model` и могут обрабатываться, как и любые другие модели, в соответствующих средах машинного обучения (МО).
-- **Простота**: В библиотеке практически не используются абстракции. "Все в одном файле" является основной концепцией: прямой проход модели полностью определяется в одном файле, так что сам код понятен, но при этом доступен для взлома.
+- **Простота использования**: Скачать, загрузить и использовать современную модель NLP для инференса можно всего в две строчки кода.
```
[the remark this answers is not in the document]
Comment (blademoon) [reply chapters/ru/chapter2/1.mdx:12]: OK.

File: chapters/ru/chapter2/2.mdx [hunk 15]
```diff
@@ -1,29 +1,29 @@
 <FrameworkSwitchCourse {fw} />
 
-# Внутри конвейера
+# За конвейером[[behind-the-pipeline]]
```
[the remark this answers is not in the document]
Comment (blademoon) [reply chapters/ru/chapter2/2.mdx:3]: OK.

File: chapters/ru/chapter2/2.mdx [hunk 16]
```diff
@@ -46,33 +46,33 @@ classifier(
...
-и на выходе получали:
+и получил:
 
 ```python out
 [{'label': 'POSITIVE', 'score': 0.9598047137260437},
  {'label': 'NEGATIVE', 'score': 0.9994558095932007}]
 ```
 
-Как мы уже увидели в [Главе 1](../chapter1/1), данный конвейер включает в себя три шага: предварительная обработка, передача входных данных через модель и постобработка:
+Как мы видели в [Главе 1] (../chapter1), этот конвейер объединяет три этапа: предобработку, пропуск входных данных через модель и постобработку:
```
[the remark this answers is not in the document]
Comment (blademoon) [reply chapters/ru/chapter2/2.mdx:56]: OK.

File: chapters/ru/chapter2/2.mdx [hunk 17]
```diff
@@ -210,24 +209,24 @@ print(outputs.last_hidden_state.shape)
 ```
 {/if}
 
-Обратите внимание, что выходные данные моделей 🤗 Transformers ведут себя как именованные кортежи или словари. Вы можете получить доступ к элементам по атрибутам (как это сделали мы) или по ключу (`outputs["last_hidden_state"]`), или даже по индексу, если вы точно знаете, где находится то, что вы ищете (`outputs[0]`).
+Обратите внимание, что выходы моделей 🤗 Transformers ведут себя как `namedtuple` или словари. Вы можете обращаться к элементам по атрибутам (как мы это делали), по ключу (`outputs["last_hidden_state"]`) или даже по индексу, если вы точно знаете, где находится искомый элемент (`outputs[0]`).
```
[the remark this answers is not in the document]
Comment (blademoon) [reply chapters/ru/chapter2/2.mdx]: aef95c177766daaefaee1c5d9cc2da821de2bbce

File: chapters/ru/chapter2/2.mdx [hunk 18]
```diff
@@ -46,33 +46,33 @@ classifier(
...
+Как мы видели в [Главе 1] (../chapter1), этот конвейер объединяет три этапа: предобработку, пропуск входных данных через модель и постобработку:
 
 <div class="flex justify-center">
-<img class="block dark:hidden" src="https://huggingface.co/datasets/huggingface-course/documentation-images/resolve/main/en/chapter2/full_nlp_pipeline.svg" alt="Полный конвейер NLP: токенизация текста, преобразование в идентификаторы и вывод с помощью модели Transformer и слоя 'head' модели."/>
-<img class="hidden dark:block" src="https://huggingface.co/datasets/huggingface-course/documentation-images/resolve/main/en/chapter2/full_nlp_pipeline-dark.svg" alt="Полный конвейер NLP: токенизация текста, преобразование в идентификаторы и вывод с помощью модели Transformer и 'head' модели."/>
+<img class="block dark:hidden" src="https://huggingface.co/datasets/huggingface-course/documentation-images/resolve/main/en/chapter2/full_nlp_pipeline.svg" alt="The full NLP pipeline: tokenization of text, conversion to IDs, and inference through the Transformer model and the model head."/>
```
[the remark this answers is not in the document]
Comment (blademoon) [reply chapters/ru/chapter2/2.mdx]: f00ba3205e89b0fd1a27d19b77118d5046acf660

File: chapters/ru/chapter2/2.mdx [hunk 19]
```diff
@@ -210,24 +209,24 @@ print(outputs.last_hidden_state.shape)
...
-### Слои "head" модели: Разбираемся в цифрах
+### Головы моделей: Извлечение смысла из чисел[[model-heads-making-sense-out-of-numbers]]
 
-Слои "head" модели принимают многомерный вектор скрытых состояний в качестве входных данных и проецируют их в другое измерение. Они обычно состоят из одного или нескольких линейных слоев:
+Головы модели принимают на вход высокоразмерный вектор скрытых состояний и проецируют его на другое измерение. Обычно они состоят из одного или нескольких линейных слоев:
 
 <div class="flex justify-center">
-<img class="block dark:hidden" src="https://huggingface.co/datasets/huggingface-course/documentation-images/resolve/main/en/chapter2/transformer_and_head.svg" alt="Нейронная сеть Transformer перед слоем 'head'."/>
-<img class="hidden dark:block" src="https://huggingface.co/datasets/huggingface-course/documentation-images/resolve/main/en/chapter2/transformer_and_head-dark.svg" alt="Нейронная сеть Transformer перед слоем с 'head'."/>
+<img class="block dark:hidden" src="https://huggingface.co/datasets/huggingface-course/documentation-images/resolve/main/en/chapter2/transformer_and_head.svg" alt="A Transformer network alongside its head."/>
```
[the remark this answers is not in the document]
Comment (blademoon) [reply chapters/ru/chapter2/2.mdx]: 8c6a58c6165eb6a6f0c73b968072f7d876716454

File: chapters/ru/chapter2/4.mdx (hunk 3 above)
Comment (blademoon) [reply chapters/ru/chapter2/4.mdx]: 3c686e5ccc1b76a9f836f4f4994578c468dcf4cc

File: chapters/ru/chapter2/4.mdx (hunk 4 above)
Comment (blademoon) [reply chapters/ru/chapter2/4.mdx]: 3c686e5ccc1b76a9f836f4f4994578c468dcf4cc

File: chapters/ru/chapter2/4.mdx (hunk 6 above)
Comment (blademoon) [reply chapters/ru/chapter2/4.mdx]: 3c686e5ccc1b76a9f836f4f4994578c468dcf4cc

File: chapters/ru/chapter2/4.mdx (hunk 7 above)
Comment (blademoon) [reply chapters/ru/chapter2/4.mdx]: 3c686e5ccc1b76a9f836f4f4994578c468dcf4cc

**Comment by blademoon** [comment]
@MKhalusova Thank you very much for your comments and edits. This will really improve the quality of the course Maria.

**Pull request merged**